### PR TITLE
refactor CheckProofOfWork rules: should fix issue #66

### PR DESCRIPTION
**Pull request by DeckerSU**
This pull request includes refactoring of the `difficulty_is_valid` and `block_is_valid_for_recent_chain` functions to make the conditions more similar to what we have in `CheckProofOfWork`, which will make them easier to understand. Additionally, it should address the chain sync issue mentioned in this link: https://github.com/KomodoPlatform/zebra/issues/66.

### CODE REVIEW

**Comment by DeckerSU** [comment]
Based on the latest changes mentioned [here](https://github.com/KomodoPlatform/zebra/issues/66#issuecomment-1692104679), it appears that the code is ready for merging. The full-sync test has also been passed. Therefore, we are now waiting for approval.

**Comment by dimxy** [comment]
With this komodo_is_special_notary_block result processing we are losing its nested error diagnosis. Maybe save the returned result and extract the error from it, if needed?

**Comment by DeckerSU** [comment]
> With this komodo_is_special_notary_block result processing we are losing its nested error diagnosis. Maybe save the returned result and extract the error from it, if needed?

You mean like this:
```rust
    let f_is_special_notary_block = match komodo_is_special_notary_block(&prepared.block, &prepared.height, network, relevant_chain.into_iter()) {
        Ok(f_is_special) => f_is_special,
        Err(e) => { // returns error if special block invalid, i.e. it should be validate as regular block
            tracing::debug!("komodo_is_special_notary_block error: {}", e.to_string());
            false
        },
    };
```
Just for printing diagnosis message?

**Comment by dimxy** [comment]
I mean in the original code komodo_is_special_notary_block() returned several variants of NotaryValidateContextError, with detailed error info, which was converted into KomodoSpecialBlockInvalid error which is further processed by zebra.
Now NotaryValidateContextError apparently is not used and KomodoSpecialBlockInvalid error simply created outside komodo_is_special_notary_block(), with no error details

**Comment by dimxy** [comment]
So I suggest saving the result and throw the error from it.
(But its up to you, I just wanted to note that the nested error is not used anymore)

**Review by dimxy — APPROVED** [review]

**Pull request merged**
